### PR TITLE
feat: add context+reducer & APIs

### PR DESCRIPTION
**Pull request by anik31**
## Added APIs for functionality & context+reducers
- create/edit/delete note
- archive/unarchive
- restore/delete-permanently
- filter notes
- create labels
- pin/unpin notes

## Files to check
- src/components/ColorPallete/ColorPallete.jsx
- src/components/CreateNote/CreateNote.jsx
- src/components/DisplayNote/HomeButtons.jsx
- src/components/DisplayNote/Note.jsx
- src/components/FilterModal/FilterModal.jsx
- src/context/archive-context.js
- src/context/note-context.js
- src/context/trash-context.js
- src/hooks/useScrollToTop.js
- src/pages/private/notes/Notes.jsx
- src/reducer/note-reducer.js
- src/utils/getFilteredNotes.js
- src/utils/getTimestamp.js
 
 _PS - Ignore other files._
 
https://user-images.githubusercontent.com/56336326/165702619-cd2f87fb-321e-45cf-9bce-21793ef73a7e.mp4

### CODE REVIEW

File: src/hooks/useScrollToTop.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { useEffect } from 'react';
+import { useLocation } from 'react-router';
+import { useNote } from '../context';
+
+export function useScrollToTop() {
```
Comment (sainath7878) [inline src/hooks/useScrollToTop.js:5]: Appreciation: Good that you've implemented this hook.

File: src/components/AddLabelModal/AddLabelModal.jsx [hunk 2]
```diff
@@ -1,13 +1,25 @@
 import "./addLabelModal.css";
+import { useState } from "react";
+import { useNote } from "../../context";
```
Comment (sainath7878) [inline src/components/AddLabelModal/AddLabelModal.jsx:3]: Suggestion: You can use absolute path

File: src/components/Label/Label.jsx [hunk 3]
```diff
@@ -1,17 +1,27 @@
...
     return (
-        <div className="label-modal">
+        <div className="label-select">
             <h4>Label</h4>
+            <i className="fas fa-times" onClick={()=>setIsLabelModalOpen(false)}></i>
             <ul>
-                <li><label><input type="checkbox"/>label1</label></li>
-                <li><label><input type="checkbox"/>label2</label></li>
+                {noteState.labels.map(label=>
+                    <li key={label}><label><input data-label={label} type="checkbox" checked={note.labels.includes(label)} onChange={addLabelToNoteHandler} />{label}</label></li>
```
Comment (sainath7878) [inline src/components/Label/Label.jsx:22]: Appreciation: Good that you've used data attribute. Didn't know about this earlier

**Review by sainath7878 — APPROVED** [review]
Everything looks good. Great Work on the UI too!

File: src/components/ColorPallete/ColorPallete.jsx [hunk 4]
```diff
@@ -1,12 +1,20 @@
...
+    const colorPalleteHandler = (e) =>{
+        setIsColorPalleteOpen(false);
+        setNote(prev=>({...prev,color:e.target.dataset.color}));
+    };
+
     return (
         <div className="color-pallete">
             {colorPallete.map(color=> 
-            <span key={color} onClick={()=>setIsColorPalleteOpen(false)} 
+            <span key={color} data-color={color} onClick={colorPalleteHandler}
```
Comment (nayyyhaa) [inline src/components/ColorPallete/ColorPallete.jsx:17]: Appreciation: good use of data attribute

File: src/pages/private/profile/Profile.jsx [hunk 5]
```diff
@@ -1,35 +1,39 @@
...
+                <h3>{user.email}</h3>
                 <button onClick={()=>logoutUser()} className="btn btn-primary-outline">LOGOUT</button>
             </div>
             <div>
                 <h4 className="text-center">Total Notes</h4>
                 <ul>
                     <li>
                         <span>Pinned : </span>
-                        <span>2</span>
+                        <span>{notes.filter(note=>note.isPinned).length}</span>
```
Comment (nayyyhaa) [inline src/pages/private/profile/Profile.jsx:24]: Liked that you are showing notes data in profile page.

**Review by nayyyhaa — APPROVED** [review]
Nice work Aniket! Code is easily understandable. Found nothing to suggest. Your UI looks great as well 💯 you can merge the changes

File: src/hooks/useScrollToTop.js (hunk 1 above)
Comment (anik31) [reply src/hooks/useScrollToTop.js:5]: Thanks

File: src/components/AddLabelModal/AddLabelModal.jsx (hunk 2 above)
Comment (anik31) [reply src/components/AddLabelModal/AddLabelModal.jsx:3]: Okay I will add it

File: src/components/Label/Label.jsx (hunk 3 above)
Comment (anik31) [reply src/components/Label/Label.jsx:22]: Thanks

File: src/components/ColorPallete/ColorPallete.jsx (hunk 4 above)
Comment (anik31) [reply src/components/ColorPallete/ColorPallete.jsx:17]: Thanks

**Pull request merged**